examples/lessons/16/app.js

```diff
@@ -46,254 +213,285 @@ export const animation = new AnimationLoop()
-        materialEmissiveColor: [0, 0, 0]
+    rttFramebuffer = new Framebuffer(gl, {
+      width: FB_WIDTH,
+      height: FB_HEIGHT,
+      attachments: {
+        [GL.COLOR_ATTACHMENT0]: rttTexture,
+        [GL.DEPTH_ATTACHMENT]: renderbuffer
       }
     });
-    box.setScale(new Vec3(2, 2, 2));
 
-    return new Group({children: [moon, box]});
-  });
-})
-
-function drawInnerScene({gl, tick}) {
-  const theta = tick * 0.04;
-
-  moon
-    .setPosition(new Vec3(
-      rho * Math.cos(theta),
-      0,
-      rho * Math.sin(theta)
-    ))
-    .updateMatrix();
-
-  box
-    .setPosition(new Vec3(
-      rho * Math.cos(Math.PI + theta),
-      0,
-      rho * Math.sin(Math.PI + theta)
-    ))
-    .updateMatrix();
-
-  gl.viewport(0, 0, screenWidth, screenHeight);
-
-  withParameters(gl, {
-    framebuffer: fb,
-    viewport: [0, 0, screenWidth, screenHeight]
-  }, () => {
-    innerScene.render({camera: innerCamera});
-  });
+    if (DISABLE_FB === true) {
+      rttFramebuffer = null;
+    }
 }
 
-function createOuterScene(gl) {
-  return Promise.all([
-    loadFiles({urls: ['macbook.json']})
-    .catch(error => window.alert('Unable to load macbook model: ' + error)),
-
-    loadProgram(gl, {
-      vs: '../../../shaderlib/render-tex.vs.glsl',
-      fs: '../../../shaderlib/render-tex.fs.glsl'
-    }),
-  ])
-  .then([macbookJSON, program] => {
-    const screenWidth = 512;
-    const screenHeight = 512;
-    const screenRatio = 1.66;
-
-    var json = JSON.parse(macbookJSON);
-    json.shininess = 5;
-    json.uniforms = {
-      'enableSpecularHighlights': true,
-      'materialAmbientColor': [1, 1, 1],
-      'materialDiffuseColor': [1, 1, 1],
-      'materialSpecularColor': [1.5, 1.5, 1.5],
-      'materialEmissiveColor': [0, 0, 0]
-    };
-
-    const macbookscreen = new Model({
-      program,
-      drawMode: GL.TRIANGLE_STRIP,
-      vertexCount: 4,
-      attributes: {
-        normals: new Float32Array([
-          0, -0.965926, 0.258819,
-          0, -0.965926, 0.258819,
-          0, -0.965926, 0.258819,
-          0, -0.965926, 0.258819
-        ]),
-        vertices: new Float32Array([
-          0.580687, 0.659, 0.813106,
-          -0.580687, 0.659, 0.813107,
-          0.580687, 0.472, 0.113121,
-          -0.580687, 0.472, 0.113121
-        ]),
-        texCoords: new Float32Array([
-          1.0, 1.0,
-          0.0, 1.0,
-          1.0, 0.0,
-          0.0, 0.0
-        ])
-      }),
-      uniforms: {
-        shininess: 0.2,
-        enableSpecularHighlights: false,
-        materialAmbientColor: [0, 0, 0],
-        materialDiffuseColor: [0, 0, 0],
-        materialSpecularColor: [0.5, 0.5, 0.5],
-        materialEmissiveColor: [1.5, 1.5, 1.5]
-      }
-    });
+function generateTextureForLaptopScreen(gl, tick, aspect, moon, cube, tSquare) {
+  moonAngle += moonAngleDelta;
+  cubeAngle += cubeAngleDelta;
+  gl.viewport(0, 0, FB_WIDTH, FB_HEIGHT);
+
+  if (!DISABLE_FB) {
+    rttFramebuffer.bind();
+  }
+
```

Comment:
> We should consider put gl.viewport() calls into Framebuffer.bind() to minimize the surprise, tracked by our state manager

Good point. I think we have mostly used FBs of the same size as the screen.

I had a crazy idea - what if we made the `draw` call as a member of the `Framebuffer` class?

const screen = Framebuffer.getDefault();
screen.draw(...);
const fb = new Framebuffer(...);
fb.draw(...); // binds buffer and sets viewport temporarily...

There would be no need to bind framebuffers or `viewport` etc. The code would make it super clear what framebuffer we were working with.

We'd need to support backward compatibility of course...